# **Exploratory Data Analysis**

## **0. Imports**

In [1]:
import numpy as np 
import pandas as pd

## **1. Load Data**

In [11]:
customer = pd.read_csv('../data/raw/raw_customers.csv')
item = pd.read_csv('../data/raw/raw_order_items.csv')
order = pd.read_csv('../data/raw/raw_orders.csv')
product = pd.read_csv('../data/raw/raw_products.csv')
store = pd.read_csv('../data/raw/raw_stores.csv')

## **2. EDA: Stores**

In [ ]:
print("Store dataset shape:", store.shape)
print("Store columns and NULL vales:")
print(store.isnull().sum())
display(store.head(10))


Store dataset shape: (10, 5)
Store columns and NULL vales:
store_name      0
city            0
address         0
postal_code     0
manager_name    0
dtype: int64


,store_name,city,address,postal_code,manager_name
0,NorthStore,Amsterdam,Hoofdstraat 48,8372 PR,Jan de Vries
1,SouthStore,rotterdam,Lijnbaan 168,5658 NT,Maria Jansen
2,EastStore,UTRECHT,Lijnbaan 2,2957 VK,Piet Bakker
3,WestStore,Den Haag,Lijnbaan 1,6651 YH,Sophie Peters
4,CenterStore,Eindhoven,Lijnbaan 5,1233 JM,Lars Visser
5,MainStore,Tilburg,Kerkstraat 175,1956 VI,Emma Mulder
6,PlazaStore,Groningen,Hoofdstraat 33,8845 SX,Klaas Smit
7,HarborStore,Breda,Hoofdstraat 6,2698 QA,Anna de Boer
8,MarketStore,NIJMEGEN,Marktstraat 172,7918 HI,Tom van Dijk
9,CityStore,Haarlem,Kerkstraat 189,7612 FG,Lisa Hendriks


### **Findings (Store Table):**
- 10 rows, 5 columns
- No missing Values
- Inconsistent city casing
- No primary keys present

## **3. EDA: Products**


In [51]:
print("Products dataset shape:", product.shape)
print("\nProducts columns and NULL values:")
print(product.isnull().sum())

print("\nTrailing spaces per column:")
for col in product.select_dtypes(include='object').columns:
    count = product[col].str.endswith(' ').sum()
    print(f"  {col}: {count}")

print("\nUnique categories:", product['category'].nunique())
display(product.head())

Products dataset shape: (145, 5)

Products columns and NULL values:
product_name       0
category           0
price             13
supplier           0
stock_quantity    16
dtype: int64

Trailing spaces per column:
  product_name: 25
  category: 0
  supplier: 0

Unique categories: 30


,product_name,category,price,supplier,stock_quantity
0,Laptop,ELECTRONICS,NaN,GlobalSupply,6.0
1,Smartphone,Electronics,945.45,TechSupply BV,171.0
2,Smartphone,ELECTRONICS,945.45,Prime Goods,171.0
3,Tablet,electronics,1346.72,Prime Goods,111.0
4,Headphones,Elektronica,750.23,GlobalSupply,471.0


### **Findings (Product Table):**
- 145 rows, 5 columns
- 29 missing values (13 in price, 16 in stock quantity)
- 25 products with trailing spaces in product_name 
- 30 unique category values for 6 actual categories:
    - Inconsistent casing: Electronics, electronics, ELECTRONICS, electronic, Elektronica
    - Dutch translations mixed in: Voedsel, Wonen, Kleding, Schoonheid, Sport
    - Plural/singular variations: Food vs Foods, Sport vs Sports
    - Alternative names: Personal Care, Home & Living, Sporting Goods, clothes
- No primary key present 

## **4. EDA: Customers**

In [50]:
print("Customer dataset shape:", customer.shape)
print("\nCustomer columns and NULL values:")
print(customer.isnull().sum())

print("\nTrailing spaces per column:")
for col in customer.select_dtypes(include='object').columns:
    count = customer[col].str.endswith(' ').sum()
    print(f"  {col}: {count}")

print("\nDuplicate rows:", customer.duplicated().sum())

print("\nUnique values per categorical column:")
for col in customer.select_dtypes(include='object').columns:
    print(f"\n{col} — {customer[col].nunique()} unique values")
    print(customer[col].value_counts().head(10))

display(customer.head())

Customer dataset shape: (640, 7)

Customer columns and NULL values:
first_name        0
last_name         0
email            77
phone            42
city              0
country           0
date_of_birth     0
dtype: int64

Trailing spaces per column:
  first_name: 0
  last_name: 0
  email: 0
  city: 0
  country: 0
  date_of_birth: 0

Duplicate rows: 40

Unique values per categorical column:

first_name — 30 unique values
first_name
Lucas       33
Mia         29
Finn        27
Benjamin    26
Laura       26
Sem         26
Ethan       24
Isabella    23
Bram        23
Noah        23
Name: count, dtype: int64

last_name — 20 unique values
last_name
van Leeuwen    50
de Groot       44
Dekker         41
de Vries       36
Bos            35
Peters         34
Mulder         33
Jansen         32
Meijer         32
Visser         32
Name: count, dtype: int64

email — 528 unique values
email
julia.vos24@gmail.com           2
noah.bakker13@gmail.com         2
tim.degroot66@hotmail.com       2
mason.de

,first_name,last_name,email,phone,city,country,date_of_birth
0,Nora,Dekker,nora.dekker87@kpn.nl,669847212.0,Haarlem,The Netherlands,1974-03-18
1,Levi,de Groot,levi.degroot62@yahoo.com,686307047.0,Haarlem,The Netherlands,1994-08-10
2,Aiden,Hendriks,aiden.hendriks15@ziggo.nl,652549479.0,Haarlem,The Netherlands,1962-10-18
3,Mason,van Leeuwen,mason.vanleeuwen89@hotmail.com,NaN,Eindhoven,nl,1986-05-07
4,Sophia,Jansen,NaN,620517177.0,Eindhoven,nederland,1990-01-07


### **Findings: (Customer Table)**
- 640 rows, 7 columns
- 77 missing emails 
- 42 missing phone numbers 
- No trailing spaces in any column
- 40 exact duplicate rows
- 7 unique country variations for 1 actual country:
    - Inconsistent casing: Netherlands, NETHERLANDS
    - Abbreviations: NL, NLD, nl
    - Alternative names: The Netherlands, nederland
- city and date_of_birth appear consistent, no cleaning needed
- Duplicate emails detected — likely caused by duplicate rows
- No primary key present — needs to be generated in cleaning

## **5. EDA: Orders**


In [52]:
print("Orders dataset shape:", order.shape)
print("\nOrders columns and NULL values:")
print(order.isnull().sum())

print("\nTrailing spaces per column:")
for col in order.select_dtypes(include='object').columns:
    count = order[col].str.endswith(' ').sum()
    print(f"  {col}: {count}")

print("\nDuplicate rows:", order.duplicated().sum())

print("\nUnique values per categorical column:")
for col in order.select_dtypes(include='object').columns:
    print(f"\n{col} — {order[col].nunique()} unique values")
    print(order[col].value_counts().head(10))

display(order.head())

Orders dataset shape: (1200, 5)

Orders columns and NULL values:
customer_name      0
order_date         0
store_name       101
status             0
total_amount       0
dtype: int64

Trailing spaces per column:
  customer_name: 0
  order_date: 0
  store_name: 0
  status: 0

Duplicate rows: 0

Unique values per categorical column:

customer_name — 343 unique values
customer_name
Benjamin van den Berg    12
Noah Jansen              12
Lucas van Dijk           12
Benjamin Bakker          11
Laura de Vries           11
Bram van Leeuwen         11
Daan Dekker              10
Liam Dekker              10
Oliver van Leeuwen       10
Mia Jansen                9
Name: count, dtype: int64

order_date — 1017 unique values
order_date
06-02-2022           4
08-27-2022           4
2022-11-27           3
12/04/2024           3
2024-01-18           3
24/01/2023           3
2023-03-08           3
2022-10-14           3
13 September 2024    3
2022-04-17           3
Name: count, dtype: int64

store_name 

,customer_name,order_date,store_name,status,total_amount
0,Nora de Groot,01/10/2022,MarketStore,Completed,750.30
1,Sara Meijer,05-05-2024,SouthStore,SHIPPED,104.10
2,Isabella Jansen,23/11/2022,CityStore,Cancelled,455.29
3,Lucas Meijer,13/07/2023,SouthStore,Completed,695.19
4,Ava Bakker,2024-01-08,PlazaStore,cancelled,728.17


### **Findings: (Order Table)**
- 1200 rows, 5 columns
- 101 missing store names 
- No trailing spaces in any column
- No duplicate rows
- 4 inconsistent date formats in order_date:
    - ISO format: 2022-11-27
    - European: 24/01/2023
    - American: 12/04/2024
    - Written out: 13 September 2024
- 10 unique status variations for 3 actual statuses:
    - Inconsistent casing: completed, Completed, COMPLETED
    - Inconsistent casing: shipped, Shipped, SHIPPED
    - Inconsistent casing: pending, Pending
    - Missing: cancelled, Cancelled
- customer_name references customer by name instead of ID — needs foreign key 
- store_name consistent, no casing issues
- No primary key present 

## **6. EDA: Order Items**


In [53]:
print("Order items dataset shape:", item.shape)
print("\nOrder items columns and NULL values:")
print(item.isnull().sum())

print("\nTrailing spaces per column:")
for col in item.select_dtypes(include='object').columns:
    count = item[col].str.endswith(' ').sum()
    print(f"  {col}: {count}")

print("\nDuplicate rows:", item.duplicated().sum())

print("\nUnique values per categorical column:")
for col in item.select_dtypes(include='object').columns:
    print(f"\n{col} — {item[col].nunique()} unique values")
    print(item[col].value_counts().head(10))

print("\nWrong line totals:")
item['expected_total'] = item['unit_price'] * item['quantity']
print(item[item['line_total'] != item['expected_total']].shape[0], "incorrect line totals")

display(item.head())

Order items dataset shape: (3000, 5)

Order items columns and NULL values:
order_reference      0
product_name         0
quantity           265
unit_price         236
line_total         481
dtype: int64

Trailing spaces per column:
  order_reference: 0
  product_name: 0

Duplicate rows: 4

Unique values per categorical column:

order_reference — 1183 unique values
order_reference
Noah Jansen              18
Benjamin van den Berg    18
Lucas van Dijk           18
Benjamin Bakker          16
Mia Bos                  15
Oliver van Leeuwen       15
Charlotte Peters         14
Daan Bakker              13
Ava de Groot             12
Finn de Groot            12
Name: count, dtype: int64

product_name — 120 unique values
product_name
Instant Noodles    44
Dried Fruit        36
Microphone         36
Lamp               35
Laundry Basket     35
Storage Box        35
Crackers           34
Lip Balm           34
Rice               33
Sweater            33
Name: count, dtype: int64

Wrong line totals

,order_reference,product_name,quantity,unit_price,line_total,expected_total
0,ORD_0860,Mouse,6.0,923.37,5540.22,5540.22
1,Evelyn Peters,Lamp,6.0,129.35,659.68,776.10
2,ORD_0007,Fitness Tracker,5.0,228.04,570.10,1140.20
3,Liam Dekker,Conditioner,2.0,38.22,76.44,76.44
4,ORD_0761,Sports Watch,9.0,NaN,NaN,NaN


### **Findings: (Order Items Table)**
- 3000 rows, 5 columns
- 265 missing quantities 
- 236 missing unit prices 
- 481 missing line totals  
- No trailing spaces in any column
- 4 duplicate rows
- 1151 incorrect line totals, unit_price * quantity does not match line_total
- order_reference inconsistent, mix of ORD_XXXX format and customer names
- product_name consistent, no casing issues
- No primary key or foreign keys present 

# **7. Summary**

| Table | Rows | Key Issues | Action |
|---|---|---|---|
| stores | 10 | Inconsistent city casing | Standardise to title case |
| products | 145 | 30 category variations, 25 trailing spaces, 13 NULL prices | Standardise categories, trim names, keep NULL prices |
| customers | 640 | 40 duplicates, 7 country variations, 77 NULL emails | Remove duplicates, standardise country to NL |
| orders | 1200 | 4 date formats, 10 status variations, 101 NULL stores | Standardise dates to YYYY-MM-DD, standardise status |
| order_items | 3000 | 1151 wrong line totals, 265 NULL quantities, 4 duplicates | Delete incomplete rows, recalculate line totals |

Full cleaning plan implemented in [04_cleaning.sql](../sql/04_cleaning.sql)
